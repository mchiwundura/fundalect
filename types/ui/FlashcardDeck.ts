import { RelativePathString } from "expo-router";
import { ColorValue } from "react-native";

export interface FlashcardDeckProps {
  title: string;
  color: ColorValue;
  link: RelativePathString;
  setDeck: any;
  completion: number;
  courseTitle?: string;
  courseId?: number;
}
