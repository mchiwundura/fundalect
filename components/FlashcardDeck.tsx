import {
  StyleSheet,
  TouchableOpacity,
  useWindowDimensions,
  View,
} from "react-native";
import React from "react";
import { ThemedText } from "./ThemedText";
import { useRouter } from "expo-router";
import Animated from "react-native-reanimated";
import { FlashcardDeckProps } from "@/types/ui/FlashcardDeck";

const FlashcardDeck = ({
  title,
  color,
  link,
  completion,
  courseTitle,
  setDeck,
}: FlashcardDeckProps) => {
  //  Hooks
  const router = useRouter();
  const { width } = useWindowDimensions();

  return (
    <TouchableOpacity
      style={[
        styles.wrapper,
        {
          width: width > 600 ? 300 : "100%",
          marginHorizontal: width > 600 ? 10 : 0,
        },
      ]}
      activeOpacity={0.85}
      onPress={() => {
        router.push(link);
        setDeck();
      }}
    >
      {/* Peeking background card */}
      <View style={[styles.cardPeek, { backgroundColor: color }]} />

      {/* Main card */}
      <Animated.View style={[styles.container, { backgroundColor: color }]}>
        <View style={styles.topRow}>
          <ThemedText style={styles.courseTitle}>{courseTitle}</ThemedText>
        </View>

        <ThemedText type="title" style={styles.title}>
          {title}
        </ThemedText>

        <View>
          <ThemedText type="subtitle" style={styles.subtitle}>
            Completion: {completion}%
          </ThemedText>

          <View style={styles.progressBarBackground}>
            <View
              style={[styles.progressBarFill, { width: `${completion}%` }]}
            />
          </View>
        </View>
      </Animated.View>
    </TouchableOpacity>
  );
};

export default FlashcardDeck;

const styles = StyleSheet.create({
  wrapper: {
    height: 210,
    position: "relative",
    marginVertical: 10,
  },
  button: {
    padding: 10,
    borderRadius: 10,
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
  cardPeek: {
    position: "absolute",
    top: -8,
    left: 8,
    right: 8,
    height: 200,
    borderRadius: 20,
    opacity: 0.3,
    zIndex: 0,
  },
  container: {
    flex: 1,
    borderRadius: 20,
    padding: 20,
    zIndex: 1,
    shadowColor: "#000",
    justifyContent: "space-between",
    shadowOffset: { width: 0, height: 6 },
    shadowOpacity: 0.2,
    shadowRadius: 8,
    elevation: 5,
  },
  topRow: {
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
  },
  courseTitle: {
    fontSize: 16,
    fontWeight: "600",
  },
  title: {
    marginTop: 15,
    fontSize: 32,
  },
  subtitle: {
    marginVertical: 10,
    opacity: 0.7,
    fontSize: 16,
  },
  progressBarBackground: {
    width: "100%",
    height: 10,
    backgroundColor: "rgba(255, 255, 255, 0.4)",
    borderRadius: 5,
    overflow: "hidden",
  },
  progressBarFill: {
    height: "100%",
    backgroundColor: "#00ff9d",
    borderRadius: 5,
  },
});
