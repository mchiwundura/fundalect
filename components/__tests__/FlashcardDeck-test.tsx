import { render } from "@testing-library/react-native";
import FlashcardDeck from "../FlashcardDeck";
import { RelativePathString } from "expo-router";

describe("Flashcard Deck Test", () => {
  it("Renders the card component", () => {
    const tree = render(
      <FlashcardDeck
        title="Pharmacology"
        color="yellow"
        completion={20}
        link={"/courses" as RelativePathString}
        setDeck={() => {}}
        courseId={1}
        courseTitle="pharmacology"
        key={2}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
