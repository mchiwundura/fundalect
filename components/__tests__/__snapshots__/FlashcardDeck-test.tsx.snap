// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flashcard Deck Test Renders the card component 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "height": 210,
      "marginHorizontal": 10,
      "marginVertical": 10,
      "opacity": 1,
      "position": "relative",
      "width": 300,
    }
  }
>
  <View
    style={
      [
        {
          "borderRadius": 20,
          "height": 200,
          "left": 8,
          "opacity": 0.3,
          "position": "absolute",
          "right": 8,
          "top": -8,
          "zIndex": 0,
        },
        {
          "backgroundColor": "yellow",
        },
      ]
    }
  />
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    jestInlineStyle={
      [
        {
          "borderRadius": 20,
          "elevation": 5,
          "flex": 1,
          "justifyContent": "space-between",
          "padding": 20,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 6,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 8,
          "zIndex": 1,
        },
        {
          "backgroundColor": "yellow",
        },
      ]
    }
    style={
      {
        "backgroundColor": "yellow",
        "borderRadius": 20,
        "elevation": 5,
        "flex": 1,
        "justifyContent": "space-between",
        "padding": 20,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 6,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 8,
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "rgba(0,0,0,0.8)",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 16,
              "fontWeight": "600",
            },
          ]
        }
      >
        pharmacology
      </Text>
    </View>
    <Text
      style={
        [
          {
            "color": "rgba(0,0,0,0.8)",
          },
          undefined,
          {
            "fontSize": 32,
            "fontWeight": "bold",
            "lineHeight": 32,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "fontSize": 32,
            "marginTop": 15,
          },
        ]
      }
    >
      Pharmacology
    </Text>
    <View>
      <Text
        style={
          [
            {
              "color": "rgba(0,0,0,0.8)",
            },
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 20,
              "fontWeight": "bold",
            },
            undefined,
            undefined,
            {
              "fontSize": 16,
              "marginVertical": 10,
              "opacity": 0.7,
            },
          ]
        }
      >
        Completion: 
        20
        %
      </Text>
      <View
        style={
          {
            "backgroundColor": "rgba(255, 255, 255, 0.4)",
            "borderRadius": 5,
            "height": 10,
            "overflow": "hidden",
            "width": "100%",
          }
        }
      >
        <View
          style={
            [
              {
                "backgroundColor": "#00ff9d",
                "borderRadius": 5,
                "height": "100%",
              },
              {
                "width": "20%",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
